.github/workflows/build-and-push-docker-image.yml

```diff
@@ -0,0 +1,55 @@
+---
+
+name: 🏗️
+
+on:   # yamllint disable-line rule:truthy
+  pull_request:
+  push:
+    branches: ["release/*", "unstable/*"]
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: Docker image tag
+        required: true
+        type: string
+
+jobs:
+  build-and-push:
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    steps:
+    - uses: actions/checkout@v4
+    - name: Build Docker image
+      run: |
+        DOCKER_TAG="${DOCKER_TAG/'/'/'-'}"
+        DOCKER_TAG_MAJOR=$(echo "$DOCKER_TAG" | cut -d '.' -f 1)
+        DOCKER_TAG_MAJOR_MINOR=$(echo "$DOCKER_TAG" | cut -d '.' -f 1-2)
+        IMAGE="ghcr.io/$GITHUB_REPOSITORY:${DOCKER_TAG}"
+        IMAGE_MAJOR="ghcr.io/$GITHUB_REPOSITORY:${DOCKER_TAG_MAJOR}"
+        IMAGE_MAJOR_MINOR="ghcr.io/$GITHUB_REPOSITORY:${DOCKER_TAG_MAJOR_MINOR}"
+        echo "IMAGE=$IMAGE" >>"$GITHUB_ENV"
+        echo "IMAGE_MAJOR=$IMAGE_MAJOR" >>"$GITHUB_ENV"
+        echo "IMAGE_MAJOR_MINOR=$IMAGE_MAJOR_MINOR" >>"$GITHUB_ENV"
+        docker build . \
+          --build-arg BUILDKIT_INLINE_CACHE=1 \
+          --cache-from $IMAGE \
+          --tag $IMAGE
+        docker tag $IMAGE $IMAGE_MAJOR
+        docker tag $IMAGE $IMAGE_MAJOR_MINOR
+      env:
+        DOCKER_TAG: ${{ inputs.tag || github.ref_name }}
+    - name: Log in to GHCR
+      if: github.event_name != 'pull_request'
+      run: >-
+        echo ${{ secrets.GITHUB_TOKEN }} |
+          docker login ghcr.io -u $GITHUB_ACTOR --password-stdin
+    - name: Push Docker image to GHCR
+      if: github.event_name != 'pull_request'
```

Comment:
Is it possible to pass the image between containers? It would be a good idea to pass it to the tests job and then to another one that would actually publish it, as opposed to publishing and then testing whether it should've been published in the first place...

Reply:
Do you mean, "pass the image between _workflow jobs_?" Not directly. A different workflow job is a different virtual machine. We could [upload and download artifacts](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/storing-and-sharing-data-from-a-workflow), but that seems like unnecessary complexity.

What about just running the smoke tests before the Docker builds?

Reply:
Yeah, I was imagining using the artifacts. In the past there was one project where I was publishing to GHCR from one job with some unique tag and then pulling from the following one. But using artifacts is better.

It's indeed more complex, but that's how I build testing these days. OTOH, since you asked to cut the scope, I'll consider this a future improvement idea for now.